### PR TITLE
[Feat] Add Github Actions as main CI

### PR DESCRIPTION
**Pull request by edgarriba**
### Description

This PR refactors our CI infrastructure and migrates to Github Actions.

Includes the following tests:
- documentation (pre-submit check + timed test)
- code style check (pre-submit check + timed test)
- code static analysis check (pre-submit check + timed test)
- python package: sanity check which downloads package from pypi (post-submit check + timed test)
- CPU tests (pre-submit check + timed test)
  - python version: [3.6, 3.7, 3.8]
  - python channel: [latest, nightly]
  - os: [linux, macos, windows]
- CPU test against torch versions (pre-submit check + timed test)
  - supported torch versions: [1.5.1, 1.6.0]
- CUDA tests (post-submit check + timed test). See [`vulnerability`](https://docs.github.com/en/free-pro-team@latest/actions/hosting-your-own-runners/about-self-hosted-runners#self-hosted-runner-security-with-public-repositories) issue.
  - python version: [3.7]
  - cuda version: [9.2, 10.1, 10.2]
  - python channel: [latest, nightly]
  - os: [linux]

Misc:
- disables CircleCI
- reorganizes requirements files
- add `@pytest.mark.xfail` tag to known tests that crash (will be fixed soon)

### Status
**Ready**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [x] Did you discuss the functionality or any breaking changes before ?
- [x] **Pass all tests**: did you test in local ? `make test`
- [x] Unittests: did you add tests for your new functionality ?
- [x] Documentations: did you build documentation ? `make build-docs`
- [x] Implementation: is your code well commented and follow conventions ? `make lint`
- [x] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

File: .github/workflows/tests_cpu_versions.yml
Comment (edgarriba) [inline .github/workflows/tests_cpu_versions.yml:11]: @shijianjian I detected some fails against 1.5.1 in augmentations that I think that can be fixed.

**Comment by edgarriba** [comment]
@pmeier @FGeri can you take a look as well ?

File: .github/workflows/tests_cuda.yml
Comment (FGeri) [inline .github/workflows/tests_cuda.yml:1]: Shouldn't we run the gpu test also on push?

File: .github/workflows/tests_cpu_versions.yml
Comment (FGeri) [inline .github/workflows/tests_cpu_versions.yml:29]: @edgarriba I don't know how many cores the CI platform provides, but in case more then one, we could run the tests parallel.

Comment (FGeri) [reply .github/workflows/tests_cpu_versions.yml:29]: According to the docs it has to so we could accelerate the tests a bit.

File: .github/workflows/docs.yml
Comment (pmeier) [inline .github/workflows/docs.yml:1]: You probably want to enable these tests only on `push`s to `master`. Otherwise this will run whenever any branch is pushed to `kornia/kornia`. This applies to all workflows in this PR.

File: .github/workflows/format.yml
Comment (pmeier) [inline .github/workflows/format.yml:1]: Especially the `format` workflow should run against PRs, shouldn't it?

File: .github/workflows/tests_cpu.yml
Comment (pmeier) [inline .github/workflows/tests_cpu.yml:11]: Not sure of the intention? Why do you want to test against the nightlies?

File: .github/workflows/tests_cpu.yml
Comment (pmeier) [inline .github/workflows/tests_cpu.yml:15]: Why test against `3.7` for macos and windows? Why not `3.6`?

File: .github/workflows/tests_cpu.yml
Comment (pmeier) [inline .github/workflows/tests_cpu.yml:37]: Nit

```suggestion
      run: pytest -v --device cpu --dtype float32,float64 --cov=kornia --cov-report xml ./test
```

File: .github/workflows/tests_cpu_versions.yml
Comment (pmeier) [reply .github/workflows/tests_cpu_versions.yml:11]: Isn't that irrelevant here? According to 

https://github.com/kornia/kornia/blob/339340b4a61153edcc2cb1467ba035c93c878c3f/setup.py#L27-L35

as of `kornia==0.4`, `torch==1.5` is no longer supported.

File: .github/workflows/tests_cuda.yml
Comment (pmeier) [inline .github/workflows/tests_cuda.yml:31]: Nit

```suggestion
      run: pytest -v --device cuda --dtype float32,float64 ./test
```

File: docs/requirements.txt
Comment (pmeier) [inline docs/requirements.txt:3]: Don't you want specific versions of `torch` and `torchvision`?

File: .github/workflows/tests_cuda.yml
Comment (FGeri) [inline .github/workflows/tests_cuda.yml:6]: According to https://docs.nvidia.com/deeplearning/frameworks/support-matrix/index.html this image is based on cuda 10.2. Not sure if the 9.2, 10.1 run in it properly.

File: .github/workflows/tests_cpu_versions.yml
Comment (edgarriba) [reply .github/workflows/tests_cpu_versions.yml:11]: @pmeier can be. It\s kind of messy the versioning right now. not sure what could be the proper way to handle it

File: .github/workflows/tests_cuda.yml
Comment (edgarriba) [reply .github/workflows/tests_cuda.yml:1]: sure. I'll fix

File: .github/workflows/tests_cpu_versions.yml
Comment (edgarriba) [reply .github/workflows/tests_cpu_versions.yml:29]: I was checking `pytest-xdist` that have this features already
https://pypi.org/project/pytest-xdist/
`pytest -n NUMCPUS`

File: .github/workflows/tests_cpu.yml
Comment (edgarriba) [reply .github/workflows/tests_cpu.yml:11]: so that we can have an idea of  our branch against future pytorch releases. Otherwise it's a surprise every time they release.

File: .github/workflows/tests_cuda.yml
Comment (edgarriba) [reply .github/workflows/tests_cuda.yml:6]: I took this from monai: https://github.com/Project-MONAI/MONAI/blob/master/.github/workflows/pythonapp.yml#L145
they seems to at least test against 10.1

File: .github/workflows/tests_cpu.yml
Comment (edgarriba) [reply .github/workflows/tests_cpu.yml:15]: no specific reason

File: docs/requirements.txt
Comment (edgarriba) [reply docs/requirements.txt:3]: maybe a lower bound >=1.6

File: .github/workflows/docs.yml
Comment (edgarriba) [reply .github/workflows/docs.yml:1]: not sure if that happens that often

File: .github/workflows/tests_cuda.yml
Comment (FGeri) [reply .github/workflows/tests_cuda.yml:6]: As far as I understand it they also use different images for different pytorch packages.

File: .github/workflows/tests_cpu_versions.yml
Comment (edgarriba) [reply .github/workflows/tests_cpu_versions.yml]: it seems to cause a crash:
`
[gw0] linux Python 3.6.12 cwd: /home/runner/work/kornia/kornia
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/usr/share/miniconda/envs/test/lib/python3.6/site-packages/_pytest/main.py", line 189, in wrap_session
INTERNALERROR>     config.hook.pytest_sessionstart(session=session)
INTERNALERROR>   File "/usr/share/miniconda/envs/test/lib/python3.6/site-packages/pluggy/hooks.py", line 286, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/usr/share/miniconda/envs/test/lib/python3.6/site-packages/pluggy/manager.py", line 93, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/usr/share/miniconda/envs/test/lib/python3.6/site-packages/pluggy/manager.py", line 87, in <lambda>`

File: docs/requirements.txt
Comment (pmeier) [reply docs/requirements.txt:3]: It should be the same that you require for installation. Otherwise you might get a different behavior.

File: .github/workflows/tests_cpu.yml
Comment (pmeier) [reply .github/workflows/tests_cpu.yml]: Given that you test `3.[6-8]` on Ubuntu, I would only test the oldest version as Python is generally backwards compatible. Otherwise you don't know if `3.6` causes problems on macOS or Windows.

File: docs/requirements.txt
Comment (edgarriba) [reply docs/requirements.txt:3]: then >= 1.6

File: .github/workflows/tests_cuda.yml
Comment (edgarriba) [reply .github/workflows/tests_cuda.yml:6]: I would say that since we use conda it works well. They use pip for installation which I don't think that resolves  well cuda dependencies

File: .github/workflows/tests_cpu.yml
Comment (pmeier) [reply .github/workflows/tests_cpu.yml:11]: I don't recommend that. The nightlies can be unstable and you might get failing PRs for reasons that you don't care about or you can't control. Furthermore, the nightlies, as the name implies, are updated daily and thus a lot of change can happen that you don't need to react to right away.

Here is what I would do: if a new PyTorch pre-release is announced, i.e. `torch==1.7.0.rc1`, you test `kornia` against this. Maybe create a `test-against-nightlies` workflow that is only triggered by pushes / PRs against a certain branch.

File: .github/workflows/docs.yml
Comment (pmeier) [reply .github/workflows/docs.yml:1]: As of writing this comment, you have ~15 branches in this repository besides `master`. I can't comment on how often they are used or updated.

File: .github/workflows/tests_cpu_versions.yml
Comment (pmeier) [reply .github/workflows/tests_cpu_versions.yml:11]: Could you elaborate how it could be relevant? As is, the requirements state that the current version of `kornia` require `torch>=1.6`. Thus, `torch==1.5` is not officially supported and there is no need test for its compatibility.

Comment (edgarriba) [reply .github/workflows/tests_cpu_versions.yml:11]: I suspect we didn't check well before

Comment (pmeier) [reply .github/workflows/tests_cpu_versions.yml:11]: In that case you (`kornia` core team) should decide if you want to support `torch==1.5`. If you want to, you are correct in checking this.

File: .github/workflows/tests_cpu.yml
Comment (edgarriba) [reply .github/workflows/tests_cpu.yml:11]: what mechanism could we have to not have to update by hand pre-release names ? that's why I chose nigthlies, which btw seems we are doing fine

File: .github/workflows/docs.yml
Comment (edgarriba) [reply .github/workflows/docs.yml:1]: this is a good reminder to make some house cleaning :)

File: .github/workflows/tests_cpu_versions.yml
Comment (edgarriba) [reply .github/workflows/tests_cpu_versions.yml:18]: /cc @shijianjian @ducha-aiki

File: .github/workflows/tests_cuda.yml
Comment (edgarriba) [reply .github/workflows/tests_cuda.yml:13]: @wyli any insight about this ?

Comment (edgarriba) [reply .github/workflows/tests_cuda.yml:13]: @wyli other proposals to improve our CI are welcomed

File: .github/workflows/tests_cpu_versions.yml
Comment (FGeri) [reply .github/workflows/tests_cpu_versions.yml:37]: Maybe with an earlier version? 1.3.x seems to work for me.

File: .github/workflows/tests_cuda.yml
Comment (FGeri) [reply .github/workflows/tests_cuda.yml:14]: Looks like that cudatoolkit contains the cuda runtime as well, so it indeed should be fine.

File: .github/workflows/tests_cpu_versions.yml
Comment (edgarriba) [reply .github/workflows/tests_cpu_versions.yml:37]: weird - but seems to double the execution time :(

**Comment by edgarriba** [comment]
done from side. From this point to now we can work on fixing tests I guess in a separated PR

File: .github/workflows/tests_cuda.yml
Comment (wyli) [reply .github/workflows/tests_cuda.yml:14]: Hi @edgarriba, in case you haven't seen this, monai is actually a bit concerned about this one: https://docs.github.com/en/free-pro-team@latest/actions/hosting-your-own-runners/about-self-hosted-runners#self-hosted-runner-security-with-public-repositories and may migrate the self-hosted pipelines to some other alternatives...

Comment (edgarriba) [reply .github/workflows/tests_cuda.yml:14]: @wyli thanks. Good to know

Comment (edgarriba) [reply .github/workflows/tests_cuda.yml:14]: @wyli I guess for now a solution for us could be to run this workflow on schedule against master once per day

File: .github/workflows/tests_cpu_versions.yml
Comment (shijianjian) [reply .github/workflows/tests_cpu_versions.yml:19]: @pmeier we are considering to have better backward compatibility. After this PR, we will try to enable 1.5.1 first.

**Comment by edgarriba** [comment]
@ducha-aiki @shijianjian approve and merge

**Review by ducha-aiki — APPROVED** [review]

**Review by shijianjian — APPROVED** [review]
I think we should only use 1.6.0 in this PR. We shall have another PR for enabling 1.5.1 for both CI and codes.

**Pull request merged**
